File: desktop-viewer/desktop-app/src/test/kotlin/dev/agentperf/desktop/CaptureArchiveServiceTest.kt
package dev.agentperf.desktop

import dev.agentperf.analysis.AnalysisReport
import dev.agentperf.analysis.Finding
import dev.agentperf.analysis.LayoutMetrics
import dev.agentperf.analysis.Severity
import dev.agentperf.application.TimelineChangeType
import dev.agentperf.application.TimelineDiff
import dev.agentperf.application.TimelineFrame
import dev.agentperf.application.TimelineNodeChange
import dev.agentperf.fixtures.SampleSnapshots
import dev.agentperf.protocol.CURRENT_PROTOCOL_VERSION
import dev.agentperf.protocol.CaptureFrameCodec
import dev.agentperf.protocol.LEGACY_WINDOW_ID
import dev.agentperf.protocol.ProtocolCodec
import dev.agentperf.protocol.ProtocolVersion
import dev.agentperf.protocol.ViewAttributes
import dev.agentperf.protocol.ViewNode
import dev.agentperf.protocol.normalizedToCurrentProtocol
import dev.agentperf.protocol.UnsupportedProtocolVersionException
import java.nio.file.Files
import java.nio.file.Path
import java.util.Base64
import org.junit.jupiter.api.Assertions.assertArrayEquals
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir

class CaptureArchiveServiceTest {
    @TempDir
    lateinit var tempDir: Path

    private val protocolCodec = ProtocolCodec(supportedMajor = 1)
    private val service = CaptureArchiveService(
        archiveCodec = CaptureArchiveCodec(),
        protocolCodec = protocolCodec,
    )

    @Test
    fun `service export then import restores snapshot screenshot and raw files`() {
        val path = tempDir.resolve("round-trip.apinspect")
        val snapshot = SampleSnapshots.dashboard
        val raw = CaptureRawArtifacts(
            zip = byteArrayOf(0x50, 0x4b, 0x03, 0x04),
            text = "raw text",
        )

        service.export(path, "0.1.2", snapshot, ONE_PIXEL_PNG, raw)
        val imported = service.import(path)

        assertEquals(snapshot.normalizedToCurrentProtocol(), imported.snapshot)
        assertArrayEquals(ONE_PIXEL_PNG, imported.screenshotPng)
        assertArrayEquals(raw.zip, requireNotNull(imported.rawArtifacts).zip)
        assertEquals(raw.text, requireNotNull(imported.rawArtifacts).text)
    }

    @Test
    fun `exports legacy snapshots with the current protocol version`() {
        val path = tempDir.resolve("legacy-upgraded.apinspect")
        val legacySnapshot = SampleSnapshots.dashboard.copy(
            protocolVersion = ProtocolVersion(1, 0),
            windows = emptyList(),
            defaultWindowId = null,
        )

        service.export(path, "0.1.3", legacySnapshot, ONE_PIXEL_PNG, rawArtifacts = null)

        val document = CaptureArchiveCodec().read(path)
        val exported = protocolCodec.decodeSnapshot(document.payload.snapshotJson)
        assertEquals(CURRENT_PROTOCOL_VERSION.major, document.metadata.protocolMajor)
        assertEquals(CURRENT_PROTOCOL_VERSION.minor, document.metadata.protocolMinor)
        assertEquals(CURRENT_PROTOCOL_VERSION, exported.protocolVersion)
        assertEquals(listOf(LEGACY_WINDOW_ID), exported.windows.map { it.id })
        assertEquals(LEGACY_WINDOW_ID, exported.defaultWindowId)
    }

    @Test
    fun `exports complex legacy snapshot after normalization exceeds transport limit`() {
        val path = tempDir.resolve("large-normalized-snapshot.apinspect")
        val largeProperty = "x".repeat(CaptureFrameCodec.MAX_SNAPSHOT_BYTES / 2)
        val root = (SampleSnapshots.dashboard.root as ViewNode).copy(
            attributes = ViewAttributes(
                rawProperties = mapOf("large-property" to largeProperty),
            ),
        )
        val snapshot = SampleSnapshots.dashboard.copy(root = root)

        service.export(path, "0.1.3", snapshot, ONE_PIXEL_PNG, rawArtifacts = null)

        val archivedJson = CaptureArchiveCodec().read(path).payload.snapshotJson
        assertTrue(
            archivedJson.toByteArray().size > CaptureFrameCodec.MAX_SNAPSHOT_BYTES,
        )
        val imported = service.import(path)
        assertEquals(
            largeProperty,
            (imported.snapshot.root as ViewNode).attributes.rawProperties["large-property"],
        )
    }

    @Test
    fun `service export then import restores persisted analysis report`() {
        val path = tempDir.resolve("report.apinspect")
        val report = AnalysisReport(
            metrics = LayoutMetrics(nodeCount = 5, maxDepth = 3, widestLevel = 2),
            findings = listOf(
                Finding(
                    ruleId = "layout.deep-hierarchy",
                    severity = Severity.WARNING,
                    nodeId = "root",
                    message = "deep",
                    arguments = mapOf("depth" to "12"),
                ),
            ),
        )

        service.export(path, "0.2.0", SampleSnapshots.dashboard, ONE_PIXEL_PNG, rawArtifacts = null, analysis = report)
        val imported = service.import(path)

        assertEquals(report, imported.analysis)
    }


    @Test
    fun `service export then import restores timeline history summaries`() {
        val path = tempDir.resolve("timeline.apinspect")
        val frame = TimelineFrame(
            index = 1,
            snapshot = SampleSnapshots.dashboard,
            screenshotPng = ONE_PIXEL_PNG,
            diffFromPrevious = TimelineDiff(
                previousCapturedAtEpochMillis = 1_000,
                currentCapturedAtEpochMillis = SampleSnapshots.dashboard.capturedAtEpochMillis,
                addedNodes = 1,
                removedNodes = 0,
                boundsChangedNodes = 2,
                changes = listOf(
                    TimelineNodeChange(
                        type = TimelineChangeType.CHANGED,
                        windowId = "window:main",
                        nodeId = "title",
                        nodeKey = "window:main:title",
                        className = "TextView",
                        changedProperties = listOf("bounds", "text"),
                    ),
                ),
            ),
        )

        service.export(
            target = path,
            producerVersion = "0.3.0",
            snapshot = SampleSnapshots.dashboard,
            screenshotPng = ONE_PIXEL_PNG,
            rawArtifacts = null,
            timelineFrames = listOf(frame),
        )
        val imported = service.import(path)

        assertEquals(1, imported.timelineFrames.size)
        val restored = imported.timelineFrames.single()
        assertEquals(1, restored.index)
        assertEquals(SampleSnapshots.dashboard.capturedAtEpochMillis, restored.capturedAtEpochMillis)
        assertEquals(1, restored.diffFromPrevious?.addedNodes)
        assertEquals(listOf("bounds", "text"), restored.diffFromPrevious?.changes?.single()?.changedProperties)
    }

    @Test
    fun `service imports a standalone screenshot png`() {
        val path = tempDir.resolve("manual.png")
        Files.write(path, ONE_PIXEL_PNG)

        val imported = service.importScreenshot(path)

        assertArrayEquals(ONE_PIXEL_PNG, imported.png)
        assertEquals(1, imported.widthPx)
        assertEquals(1, imported.heightPx)
    }

    @Test
    fun `standalone imported screenshot is exported as archive screenshot`() {
        val screenshotPath = tempDir.resolve("manual.png")
        val archivePath = tempDir.resolve("manual-image.apinspect")
        Files.write(screenshotPath, TWO_BY_ONE_PNG)
        val screenshot = service.importScreenshot(screenshotPath)

        service.export(
            target = archivePath,
            producerVersion = "0.3.0",
            snapshot = SampleSnapshots.dashboard,
            screenshotPng = screenshot.png,
            rawArtifacts = null,
        )

        val importedArchive = service.import(archivePath)
        assertArrayEquals(TWO_BY_ONE_PNG, importedArchive.screenshotPng)
    }

    @Test
    fun `service rejects an invalid standalone screenshot`() {
        val path = tempDir.resolve("manual.png")
        Files.write(path, byteArrayOf(1, 2, 3))

        assertThrows(CaptureArchiveFormatException::class.java) {
            service.importScreenshot(path)
        }
    }

    @Test
    fun `service rejects an invalid screenshot`() {
        val path = tempDir.resolve("invalid-png.apinspect")
        CaptureArchiveCodec().write(
            path,
            metadata(SampleSnapshots.dashboard.protocolVersion),
            CaptureArchivePayload(
                snapshotJson = protocolCodec.encodeSnapshot(SampleSnapshots.dashboard),
                screenshotPng = byteArrayOf(1, 2, 3),
            ),
        )

        assertThrows(CaptureArchiveFormatException::class.java) {
            service.import(path)
        }
    }

    @Test
    fun `service retains protocol compatibility checks`() {
        val path = tempDir.resolve("unsupported-protocol.apinspect")
        val unsupported = SampleSnapshots.dashboard.copy(
            protocolVersion = ProtocolVersion(2, 0),
        )
        CaptureArchiveCodec().write(
            path,
            metadata(unsupported.protocolVersion),
            CaptureArchivePayload(
                snapshotJson = protocolCodec.encodeSnapshot(unsupported),
                screenshotPng = ONE_PIXEL_PNG,
            ),
        )

        assertThrows(UnsupportedProtocolVersionException::class.java) {
            service.import(path)
        }
    }

    private fun metadata(version: ProtocolVersion) = CaptureArchiveMetadata(
        producerVersion = "0.1.2",
        packageName = SampleSnapshots.dashboard.packageName,
        capturedAtEpochMillis = SampleSnapshots.dashboard.capturedAtEpochMillis,
        protocolMajor = version.major,
        protocolMinor = version.minor,
    )

    private companion object {
        val ONE_PIXEL_PNG: ByteArray = Base64.getDecoder().decode(
            "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNk+A8AAQUBAScY42YAAAAASUVORK5CYII=",
        )
        val TWO_BY_ONE_PNG: ByteArray = Base64.getDecoder().decode(
            "iVBORw0KGgoAAAANSUhEUgAAAAIAAAABCAYAAAD0In+KAAAADklEQVR4nGP4z8DwH4QBEfcD/ePF9e8AAAAASUVORK5CYII=",
        )
    }
}

File: electron-viewer/apps/desktop/src/main/capture-archive-service.test.ts
import { copyFile, mkdir, mkdtemp } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { describe, expect, it } from 'vitest';
import type { LayoutSnapshot } from '@aps/layout-inspector';
import { LayoutCaptureStore } from './layout-capture-store.js';
import { CaptureArchiveService } from './capture-archive-service.js';
import { CaptureArchiveCodec } from './capture-archive-codec.js';

const ONE_PIXEL_PNG = Buffer.from('iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNk+A8AAQUBAScY42YAAAAASUVORK5CYII=', 'base64');

const KOTLIN_CAPTURE_ARCHIVE_FIXTURE = fileURLToPath(
  new URL('./fixtures/kotlin-capture.apinspect', import.meta.url),
);

const ELECTRON_CAPTURE_ARCHIVE_FIXTURE_PATH = process.env['APS_ELECTRON_CAPTURE_ARCHIVE_FIXTURE_PATH'];

function snapshot(time: number): LayoutSnapshot {
  const root = { type: 'view' as const, id: 'root', className: 'android.view.View', bounds: { left: 0, top: 0, right: 1, bottom: 1 }, visible: true, alpha: 1, children: [], attributes: { rawProperties: {} } };
  return {
    protocolVersion: { major: 1, minor: 1 }, packageName: 'com.example', capturedAtEpochMillis: time,
    display: { widthPx: 1, heightPx: 1, density: 1 }, capabilities: { viewHierarchy: true, composeSemantics: false, screenshots: true, timeline: false },
    root, windows: [{ id: 'window:1', title: 'Example', type: 'ACTIVITY', bounds: root.bounds, root }], defaultWindowId: 'window:1',
  };
}

function countNodes(): number { return 1; }

async function service(directory: string): Promise<{ service: CaptureArchiveService; store: LayoutCaptureStore }> {
  const store = new LayoutCaptureStore(join(directory, 'captures'), { countNodes });
  return { service: new CaptureArchiveService({ store, producerVersion: () => '0.4.6' }), store };
}

describe('CaptureArchiveService', () => {
  it('exports a stored capture and imports it into another store', async () => {
    const directory = await mkdtemp(join(tmpdir(), 'aps-archive-service-'));
    const source = await service(join(directory, 'source'));
    const rawArtifacts = { zip: Buffer.from([0x50, 0x4b, 3, 4]), text: 'VISIBLE WINDOW VIEW DUMP\n' };
    const archivePayloads = {
      analysisReportJson: '{"metrics":{"nodeCount":1,"maxDepth":1,"widestLevel":1},"findings":[]}',
      aiAnalysisReportJson: '{"model":"gpt-test","summary":"archive report","findings":[]}',
      timelineHistoryJson: '{"frames":[{"index":0,"capturedAtEpochMillis":1234}]}',
    };
    await source.store.add(snapshot(1234), ONE_PIXEL_PNG, '{"compose":true}', rawArtifacts, archivePayloads);
    const archive = join(directory, 'capture.apinspect');
    await source.service.export('1234', archive);

    const destination = await service(join(directory, 'destination'));
    const imported = await destination.service.import(archive);
    expect(imported.archiveVersion).toBe(2);
    expect(imported.snapshot.packageName).toBe('com.example');
    expect(await destination.store.loadScreenshot('1234')).toEqual(ONE_PIXEL_PNG);
    expect(await destination.store.loadComposeInspection('1234')).toBe('{"compose":true}');
    expect(await destination.store.loadRawArtifacts('1234')).toEqual(rawArtifacts);
    expect(await destination.store.loadArchivePayloads('1234')).toEqual(archivePayloads);
    expect(imported.composeInspectionJson).toBe('{"compose":true}');
    expect(imported.analysisReportJson).toBe(archivePayloads.analysisReportJson);
    expect(imported.aiAnalysisReportJson).toBe(archivePayloads.aiAnalysisReportJson);
    expect(imported.timelineHistoryJson).toBe(archivePayloads.timelineHistoryJson);
  });

  /**
   * The resulting fixture is consumed by Kotlin's CaptureArchiveService.
   * Regenerate it with:
   *   APS_ELECTRON_CAPTURE_ARCHIVE_FIXTURE_PATH=../desktop-viewer/layout-inspector/presentation/src/test/resources/electron-capture.apinspect \
   *     corepack pnpm@12.3.4 --dir electron-viewer exec vitest run \
   *       apps/desktop/src/main/capture-archive-service.test.ts -t "Electron-created v2 archive"
   */
  it('writes a full Electron-created v2 archive for Kotlin', async () => {
    const directory = await mkdtemp(join(tmpdir(), 'aps-electron-archive-'));
    const archive = electronCaptureArchiveFixturePath(directory);
    await mkdir(dirname(archive), { recursive: true });
    const source = await service(join(directory, 'source'));
    const capturedAtEpochMillis = 1_750_000_001_000;
    const rawArtifacts = { zip: Buffer.from([0x50, 0x4b, 3, 4]), text: 'Electron visible-window hierarchy\n' };
    const archivePayloads = {
      analysisReportJson: '{"metrics":{"nodeCount":1,"maxDepth":1,"widestLevel":1},"findings":[{"ruleId":"layout.flat","severity":"INFO","nodeId":"root","message":"Electron archive finding","arguments":{}}]}',
      aiAnalysisReportJson: '{"model":"electron-archive-model","summary":"Electron archive AI summary","findings":[{"ruleId":"ai.layout","severity":"WARNING","nodeId":"root","title":"Electron archive AI finding","message":"Review the hierarchy.","recommendation":"Keep the hierarchy shallow.","confidence":0.8,"performanceEvidenceIds":[],"sourceCandidateIds":[]}],"provenance":null}',
      timelineHistoryJson: `{"frames":[{"index":1,"capturedAtEpochMillis":${capturedAtEpochMillis},"diffFromPrevious":{"previousCapturedAtEpochMillis":${capturedAtEpochMillis - 1_000},"currentCapturedAtEpochMillis":${capturedAtEpochMillis},"addedNodes":1,"removedNodes":0,"boundsChangedNodes":1,"changes":[{"type":"CHANGED","windowId":"window:1","nodeId":"root","nodeKey":"window:1/root","className":"android.view.View","changedProperties":["bounds"]}]}}]}`,
    };
    const composeInspectionJson = JSON.stringify({
      schemaVersion: 1,
      packageName: 'com.example',
      capturedAtEpochMillis,
      frame: { frameId: `com.example:${capturedAtEpochMillis}:1`, generation: 1, mode: 'FULL', capabilities: [], roots: [] },
    });
    await source.store.add(snapshot(capturedAtEpochMillis), ONE_PIXEL_PNG, composeInspectionJson, rawArtifacts, archivePayloads);
    await source.service.export(String(capturedAtEpochMillis), archive);

    const document = await new CaptureArchiveCodec().read(archive);
    expect(document).toMatchObject({
      archiveVersion: 2,
      metadata: {
        producerVersion: '0.4.6', packageName: 'com.example', capturedAtEpochMillis,
        protocolMajor: 1, protocolMinor: 1,
      },
      payload: { screenshotPng: ONE_PIXEL_PNG, composeInspectionJson, rawArtifacts, ...archivePayloads },
    });
  });

function electronCaptureArchiveFixturePath(directory: string): string {
  if (ELECTRON_CAPTURE_ARCHIVE_FIXTURE_PATH !== undefined && ELECTRON_CAPTURE_ARCHIVE_FIXTURE_PATH.trim().length > 0) {
    return ELECTRON_CAPTURE_ARCHIVE_FIXTURE_PATH;
  }
  return join(directory, 'electron-capture.apinspect');
}

  /**
   * The archive is created by Kotlin's CaptureArchiveService and all of its
   * serializers. Regenerate it with:
   *   cd desktop-viewer && ./gradlew :layout-inspector:presentation:writeElectronInteropFixture \
   *     -PfixturePath=../electron-viewer/apps/desktop/src/main/fixtures/kotlin-capture.apinspect
   */
  it('imports a Kotlin-created v2 archive with every supported payload', async () => {
    const directory = await mkdtemp(join(tmpdir(), 'aps-kotlin-archive-'));
    const archive = join(directory, 'kotlin-capture.apinspect');
    await copyFile(KOTLIN_CAPTURE_ARCHIVE_FIXTURE, archive);

    const document = await new CaptureArchiveCodec().read(archive);
    expect(document.archiveVersion).toBe(2);
    expect(document.metadata).toEqual({
      producerVersion: 'kotlin-capture-fixture-v1',
      packageName: 'com.androidperformancestudio.sample',
      capturedAtEpochMillis: 1_750_000_000_000,
      protocolMajor: 1,
      protocolMinor: 1,
    });
    expect(document.payload.screenshotPng).toEqual(ONE_PIXEL_PNG);
    expect(document.payload.rawArtifacts).toEqual({
      zip: Buffer.from([0x50, 0x4b, 3, 4]),
      text: 'Kotlin visible-window hierarchy\n',
    });
    expect(JSON.parse(document.payload.composeInspectionJson ?? '{}')).toMatchObject({
      schemaVersion: 1,
      packageName: 'com.androidperformancestudio.sample',
      capturedAtEpochMillis: 1_750_000_000_000,
      privacy: 'SAFE_REDACTED',
      frame: { roots: [{ viewId: 7, nodes: [{ id: 1, name: 'Content' }] }] },
    });
    expect(JSON.parse(document.payload.composeInspectionJson ?? '{}').frame.details['1'].parameters[0].value).toBe('<redacted>');
    expect(JSON.parse(document.payload.analysisReportJson ?? '{}')).toMatchObject({
      metrics: { nodeCount: 3, maxDepth: 2, widestLevel: 2 },
      findings: [{ ruleId: 'layout.deep-hierarchy', severity: 'WARNING', nodeId: 'root' }],
    });
    expect(JSON.parse(document.payload.aiAnalysisReportJson ?? '{}')).toMatchObject({
      model: 'kotlin-archive-model',
      summary: 'Kotlin archive AI summary',
      findings: [{ ruleId: 'ai.layout', severity: 'ERROR', nodeId: 'root', confidence: 0.9 }],
    });
    expect(JSON.parse(document.payload.timelineHistoryJson ?? '{}')).toMatchObject({
      frames: [{
        index: 1,
        capturedAtEpochMillis: 1_750_000_000_000,
        diffFromPrevious: { addedNodes: 1, removedNodes: 0, boundsChangedNodes: 2 },
      }],
    });

    const destination = await service(join(directory, 'destination'));
    const imported = await destination.service.import(archive);
    expect(imported).toMatchObject({
      archiveVersion: 2,
      snapshot: {
        packageName: 'com.androidperformancestudio.sample',
        capturedAtEpochMillis: 1_750_000_000_000,
        display: { widthPx: 1, heightPx: 1 },
      },
      screenshotBase64: ONE_PIXEL_PNG.toString('base64'),
      composeInspectionJson: document.payload.composeInspectionJson,
      analysisReportJson: document.payload.analysisReportJson,
      aiAnalysisReportJson: document.payload.aiAnalysisReportJson,
      timelineHistoryJson: document.payload.timelineHistoryJson,
    });
    expect(await destination.store.loadScreenshot('1750000000000')).toEqual(ONE_PIXEL_PNG);
    expect(await destination.store.loadComposeInspection('1750000000000')).toBe(document.payload.composeInspectionJson);
    expect(await destination.store.loadRawArtifacts('1750000000000')).toEqual(document.payload.rawArtifacts);
    expect(await destination.store.loadArchivePayloads('1750000000000')).toEqual({
      analysisReportJson: document.payload.analysisReportJson,
      aiAnalysisReportJson: document.payload.aiAnalysisReportJson,
      timelineHistoryJson: document.payload.timelineHistoryJson,
    });
  });

  it('fails closed before exporting a persisted optional payload that is not Kotlin-compatible', async () => {
    const directory = await mkdtemp(join(tmpdir(), 'aps-archive-service-invalid-'));
    const source = await service(join(directory, 'source'));
    await source.store.add(snapshot(5678), ONE_PIXEL_PNG, undefined, undefined, {
      analysisReportJson: '{"metrics":{},"findings":[]}',
    });
    const archive = join(directory, 'invalid-optional.apinspect');
    await expect(source.service.export('5678', archive)).rejects.toThrow(/analysis metrics/);
  });
});
